### PR TITLE
OpenSslEngine return NEED_WRAP if the destination buffered filled

### LINKED ISSUE #6796: Recent commit breaks Mutual TLS
Commit f20063d26b640cb4121e092b5a3a0bb4c0a10eb5 appears to break Mutual TLS in the sense that the connection just hangs. There're no Exceptions as far as I can tell and the only failure I'm getting is when the events in [SslHandler#channelInactive(...)](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L898) fire.

Interestingly, the problem occurs only in our production environment (which has more load, latencies and RTTs between servers). The problem doesn't occur on my dev workstation (i.e. unable to reproduce in the form of a unit test).

Reverting the commit fixes the problem for me.

### Netty version

4.1.12.Final-SNAPSHOT w/ TCN 2.0.2.Final-SNAPSHOT (openssl-static)

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
If the destination buffer is completely filled during a call to OpenSslEngine#wrap(..) we may return NEED_UNWRAP because there is no data pending in the SSL buffers. However during a handshake if the SSL buffers were just drained, and filled up the destination buffer it is possible OpenSSL may produce more data on the next call to SSL_write. This means we should keep trying to call SSL_write as long as the destination buffer is filled and only return NEED_UNWRAP when the destination buffer is not full and there is no data pending in OpenSSL's buffers.

Modifications:
- If the handshake produces data in OpenSslEngine#wrap(..) we should return NEED_WRAP if the destination buffer is completely filled

Result:
OpenSslEngine returns the correct handshake status from wrap().
Fixes #6796.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 1]
```diff
@@ -779,8 +780,13 @@ private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws
                     // If this is not enough we will increase the buffer in wrap(...).
                     out = allocateOutNetBuf(ctx, 2048, 1);
                 }
+
                 SSLEngineResult result = wrap(alloc, engine, Unpooled.EMPTY_BUFFER, out);
 
+                if (previousNeedUnwrap) {
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: changes to SslHandler are just to support debugging and will be removed if this PR is accepted.

**Comment by Scottmitch** [comment]
@rkapsi - would you mind trying this PR with OPENSSL and report back the same log statements you did as in https://github.com/netty/netty/issues/6796#issuecomment-305486288?

**Comment by rkapsi** [comment]
@Scottmitch - will do first thing in the morning tomorrow

**Comment by Scottmitch** [comment]
@rkapsi verified this patch works as expected in https://github.com/netty/netty/issues/6796#issuecomment-305766517

**Comment by Scottmitch** [comment]
@rkapsi - Thanks for verifying! I reverted the debug code ... would you mind verifying that https://github.com/netty/netty/issues/6796 is fixed without any changes to `SslHandler`?

**Comment by Scottmitch** [comment]
4.1 (24f801c7d) 4.0 (77bc0f2f3)

**Comment by Scottmitch** [comment]
I will pull this in now as even if it doesn't fix #6796 it is still an issue, and the behavior from the log statements provided in https://github.com/netty/netty/issues/6796#issuecomment-305766517 demonstrates the correct behavior.

**Comment by rkapsi** [comment]
@Scottmitch running 24f801c7d1d1e66193d6ff0da3b93ce427b2b329 built from HEAD of 4.1 now. Looks good, no issues. :100:

**Pull request closed without merging**